### DISCUSSION TITLE
Auth::user return type porblem with Intelephense

### DISCUSSION THREAD

**Discussion by hfdzz** [codeigniter4/shield#1116] [Q&A] [answered]
`Auth::user()` method causes warning with Intelephense if you have own (extending) UserEntity due to type error and prevent typehinting. Is it possible to change the return type without changing the system file directly?

**Comment by kenjis** [comment]
Don't you extend `CodeIgniter\Shield\Entities\User`?

**Reply by hfdzz** [reply]
Yes. Everything works fine but a minor inconvenience. I change the return type on the PHPDoc as work around.

To clarify, it's an error not a warning, Intelephense gives `Undefined method 'myUserMethod'.intelephense(P1013)` Error on custom UserEntity method. Maybe it's just the intelephense? (PHP Intelephense for vscode)

**Reply by kenjis** [reply, marked as answer]
I got your situation.
Intelephense is no problem. Because `Auth::user()` return type is `User` not your extended UserEntity.
`User` does not have the `myUserMethod()` method.

1. Extend the Auth class and change the return type,  and use it in your app.
2. Specify the return value type of `Auth::user()` manually.
```php
/** @var UserEntity $user */
$user = Auth::user();
```

**Reply by sanchawebo** [reply]
I have the same problem with PHPStan. Could you elaborate on how to extend the `Auth` class?

For now i have this in place:
```yaml
ignoreErrors:
    - '#^Call to an undefined method CodeIgniter\\Shield\\Entities\\User\:\:getGesellschafterId\(\)\.$#'
```

**Reply by hfdzz** [reply]
You create your own `Auth` class which extend `Shield\Auth` class then you specify the `Auth::user()` return type as kenjis said.
> ```
> /** @var UserEntity $user */
> $user = Auth::user();
> ```
but then you have to change all the auth class you use, i believe including one from helper function, so i end up just ignoring the error

**Reply by kenjis** [reply]
Select 1. or 2. You don't need to do both.
